File: web/documentserver-example/csharp-mvc/Models/FileModel.cs
/**
 *
 * (c) Copyright Ascensio System SIA 2023
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

using System;
using System.Collections.Generic;
using System.IO;
using System.Web;
using System.Web.Configuration;
using System.Web.Mvc;
using System.Web.Script.Serialization;
using OnlineEditorsExampleMVC.Helpers;

namespace OnlineEditorsExampleMVC.Models
{
    // create file model
    public class FileModel
    {
        public string Mode { get; set; }  // editor mode
        public string Type { get; set; }  // editor type
        public bool IsEnabledDirectUrl { get; set; }  // is enabled direct url

        // get file url for Document Server
        public string FileUri
        {
            get { return DocManagerHelper.GetFileUri(FileName, true); }
        }

        // get file url for user
        public string FileUriUser
        {
            get { return Path.IsPathRooted(WebConfigurationManager.AppSettings["storage-path"]) ? DownloadUrl + "&dmode=emb" : DocManagerHelper.GetFileUri(FileName, false); }
        }

        public string FileName { get; set; }  // file name

        // get document type
        public string DocumentType
        {
            get { return FileUtility.GetFileType(FileName).ToString().ToLower(); }
        }

        // get document key
        public string Key
        {
            get { return ServiceConverter.GenerateRevisionId(DocManagerHelper.CurUserHostAddress() + "/" + FileName + "/" + File.GetLastWriteTime(DocManagerHelper.StoragePath(FileName, null)).GetHashCode()); }
        }

        // get the callback url
        public string CallbackUrl
        {
            get { return DocManagerHelper.GetCallback(FileName); }
        }

        public string DownloadUrl
        {
            get { return DocManagerHelper.GetDownloadUrl(FileName); }
        }

        // get the document config
        public string GetDocConfig(HttpRequest request, UrlHelper url)
        {
            var jss = new JavaScriptSerializer();

            var ext = Path.GetExtension(FileName).ToLower();  // get file extension
            var editorsMode = Mode ?? "edit";  // get editor mode

            var canEdit = DocManagerHelper.EditedExts.Contains(ext);  // check if the file with such an extension can be edited

            var id = request.Cookies.GetOrDefault("uid", null);
            var user = Users.getUser(id);  // get the user
            
            if ((!canEdit && editorsMode.Equals("edit") || editorsMode.Equals("fillForms")) && DocManagerHelper.FillFormExts.Contains(ext)) {
                editorsMode = "fillForms";
                canEdit = true;
            }
            var submitForm = editorsMode.Equals("fillForms") && id.Equals("uid-1") && false;  // check if the Submit form button is displayed or not
            var mode = canEdit && editorsMode != "view" ? "edit" : "view";  // set the mode parameter: change it to view if the document can't be edited

            // favorite icon state
            bool? favorite = user.favorite;

            var actionLink = request.GetOrDefault("actionLink", null);  // get the action link (comment or bookmark) if it exists
            var actionData = string.IsNullOrEmpty(actionLink) ? null : jss.DeserializeObject(actionLink);  // get action data for the action link

            var directUrl = DocManagerHelper.GetDownloadUrl(FileName, false);
            var createUrl = DocManagerHelper.GetCreateUrl(FileUtility.GetFileType(FileName));
            var templatesImageUrl = DocManagerHelper.GetTemplateImageUrl(FileUtility.GetFileType(FileName)); // image url for templates
            var templates = new List<Dictionary<string, string>>
            {
                new Dictionary<string, string>()
                {
                    { "image", "" },
                    { "title", "Blank" },
                    { "url", createUrl },
                },
                new Dictionary<string, string>()
                {
                    { "image", templatesImageUrl },
                    { "title", "With sample content" },
                    { "url", createUrl + "&sample=true" },
                }
            };

            // specify the document config
            var config = new Dictionary<string, object>
                {
                    { "type", Type ?? "desktop" },
                    { "documentType", DocumentType },
                    {
                        "document", new Dictionary<string, object>
                            {
                                { "title", FileName },
                                { "url", DownloadUrl },
                                { "directUrl", IsEnabledDirectUrl ? directUrl : "" },
                                { "fileType", ext.Trim('.') },
                                { "key", Key },
                                {
                                    "info", new Dictionary<string, object>
                                        {
                                            { "owner", "Me" },
                                            { "uploaded", DateTime.Now.ToShortDateString() },
                                            { "favorite", favorite}
                                        }
                                },
                                {
                                    "referenceData", new Dictionary<string, string>()
                                    {
                                        { "fileKey", !user.id.Equals("uid-0") ?
                                            jss.Serialize(new Dictionary<string, object>{
                                                {"fileName", FileName},
                                                {"userAddress", HttpUtility.UrlEncode(DocManagerHelper.CurUserHostAddress(HttpContext.Current.Request.UserHostAddress))}
                                        }) : null },
                                        {"instanceId", DocManagerHelper.GetServerUrl(false) }
                                    }
                                },
                                {
                                    // the permission for the document to be edited and downloaded or not
                                    "permissions", new Dictionary<string, object>
                                        {
                                            { "comment", editorsMode != "view" && editorsMode != "fillForms" && editorsMode != "embedded" && editorsMode != "blockcontent" },
                                            { "copy", !user.deniedPermissions.Contains("copy") },
                                            { "download", !user.deniedPermissions.Contains("download") },
                                            { "edit", canEdit && (editorsMode == "edit" || editorsMode == "view" || editorsMode == "filter" || editorsMode == "blockcontent") },
                                            { "print", !user.deniedPermissions.Contains("print") },
                                            { "fillForms", editorsMode != "view" && editorsMode != "comment" && editorsMode != "embedded" && editorsMode != "blockcontent" },
                                            { "modifyFilter", editorsMode != "filter" },
                                            { "modifyContentControl", editorsMode != "blockcontent" },
                                            { "review", canEdit && (editorsMode == "edit" || editorsMode == "review") },
                                            { "chat", !user.id.Equals("uid-0") },
                                            { "reviewGroups", user.reviewGroups },
                                            { "commentGroups", user.commentGroups },
                                            { "userInfoGroups", user.userInfoGroups },
                                            { "protect", !user.deniedPermissions.Contains("protect") }
                                        }
                                }
                            }
                    },
                    {
                        "editorConfig", new Dictionary<string, object>
                            {
                                { "actionLink", actionData },
                                { "mode", mode },
                                { "lang", request.Cookies.GetOrDefault("ulang", "en") },
                                { "callbackUrl", CallbackUrl },  // absolute URL to the document storage service
                                { "coEditing", editorsMode == "view" && user.id.Equals("uid-0") ? 
                                    new Dictionary<string, object>{
                                        {"mode", "strict"},
                                        {"change", false}
                                    } : null },
                                { "createUrl", !user.id.Equals("uid-0") ? createUrl : null },
                                { "templates", user.templates ? templates : null },
                                {
                                    // the user currently viewing or editing the document
                                    "user", new Dictionary<string, object>
                                        {
                                            { "id", !user.id.Equals("uid-0") ? user.id : null  },
                                            { "name", user.name },
                                            { "group", user.group }
                                        }
                                },
                                {
                                    // the parameters for the embedded document type
                                    "embedded", new Dictionary<string, object>
                                        {
                                            { "saveUrl", directUrl },  // the absolute URL that will allow the document to be saved onto the user personal computer
                                            { "embedUrl", directUrl },  // the absolute URL to the document serving as a source file for the document embedded into the web page
                                            { "shareUrl", directUrl },  // the absolute URL that will allow other users to share this document
                                            { "toolbarDocked", "top" }  // the place for the embedded viewer toolbar (top or bottom)
                                        }
                                },
                                {
                                    // the parameters for the editor interface
                                    "customization", new Dictionary<string, object>
                                        {
                                            { "about", true },  // the About section display
                                            { "comments", true },
                                            { "feedback", true },  // the Feedback & Support menu button display
                                            { "forcesave", false },  // adds the request for the forced file saving to the callback handler
                                            { "submitForm", submitForm },  // if the Submit form button is displayed or not
                                            {
                                                "goback", new Dictionary<string, object>  // settings for the Open file location menu button and upper right corner button
                                                    {
                                                        { "url", DocManagerHelper.GetServerUrl(false) }  // the absolute URL to the website address which will be opened when clicking the Open file location menu button
                                                    }
                                            }
                                        }
                                }
                            }
                    }
                };

            // if the secret key to generate token exists
            if (JwtManager.Enabled)
            {
                // encode the document config into a token
                var token = JwtManager.Encode(config);
                config.Add("token", token);
            }

            return jss.Serialize(config);
        }

        // get a document which will be compared with the current document
        public void GetDocumentData(out string compareConfig)
        {
            var jss = new JavaScriptSerializer();

            // get the path to the compared file
            var compareFileUrl = new UriBuilder(DocManagerHelper.GetServerUrl(true))
            {
                Path = HttpRuntime.AppDomainAppVirtualPath
                    + (HttpRuntime.AppDomainAppVirtualPath.EndsWith("/") ? "" : "/")
                    + "webeditor.ashx",
                Query = "type=assets&fileName=" + HttpUtility.UrlEncode("sample.docx")
            };

            var directCompareFileUrl = new UriBuilder(DocManagerHelper.GetServerUrl(false))
            {
                Path = HttpRuntime.AppDomainAppVirtualPath
                    + (HttpRuntime.AppDomainAppVirtualPath.EndsWith("/") ? "" : "/")
                    + "webeditor.ashx",
                Query = "type=assets&fileName=" + HttpUtility.UrlEncode("sample.docx")
            };

            // create an object with the information about the compared file
            var dataCompareFile = new Dictionary<string, object>
            {
                { "fileType", "docx" },
                { "url", compareFileUrl.ToString() }
            };

            if (IsEnabledDirectUrl)
            {
                dataCompareFile.Add("directUrl", directCompareFileUrl.ToString());
            }

            if (JwtManager.Enabled)  // if the secret key to generate token exists
            {
                var compareFileToken = JwtManager.Encode(dataCompareFile);  // encode the dataCompareFile object into the token
                dataCompareFile.Add("token", compareFileToken);  // and add it to the dataCompareFile object
            }

            compareConfig = jss.Serialize(dataCompareFile);
        }

        // get a logo config
        public void GetLogoConfig(out string logoUrl)
        {
            var jss = new JavaScriptSerializer();

            // get the path to the logo image
            var mailMergeUrl = new UriBuilder(DocManagerHelper.GetServerUrl(true))
            {
                Path = HttpRuntime.AppDomainAppVirtualPath
                    + (HttpRuntime.AppDomainAppVirtualPath.EndsWith("/") ? "" : "/")
                    + "Content\\images\\logo.png"
            };

            var directMailMergeUrl = new UriBuilder(DocManagerHelper.GetServerUrl(false))
            {
                Path = HttpRuntime.AppDomainAppVirtualPath
                    + (HttpRuntime.AppDomainAppVirtualPath.EndsWith("/") ? "" : "/")
                    + "Content\\images\\logo.png"
            };

            // create a logo config
            var logoConfig = new Dictionary<string, object>
            {
                { "fileType", "png"},
                { "url", mailMergeUrl.ToString()}
            };

            if (IsEnabledDirectUrl)
            {
                logoConfig.Add("directUrl", directMailMergeUrl.ToString());
            }

            if (JwtManager.Enabled)  // if the secret key to generate token exists
            {
                var token = JwtManager.Encode(logoConfig);  // encode logoConfig into the token
                logoConfig.Add("token", token);  // and add it to the logo config
            }

            logoUrl = jss.Serialize(logoConfig).Replace("{", "").Replace("}", "");
        }

        // get a mail merge config
        public void GetSpreadsheetConfig(out string dataSpreadsheet)
        {
            var jss = new JavaScriptSerializer();

            // get the path to the recipients data for mail merging
            var mailMergeUrl = new UriBuilder(DocManagerHelper.GetServerUrl(true))
            {
                Path =
                    HttpRuntime.AppDomainAppVirtualPath
                    + (HttpRuntime.AppDomainAppVirtualPath.EndsWith("/") ? "" : "/")
                    + "webeditor.ashx",
                Query = "type=csv"
            };

            var directMailMergeUrl = new UriBuilder(DocManagerHelper.GetServerUrl(false))
            {
                Path =
                    HttpRuntime.AppDomainAppVirtualPath
                    + (HttpRuntime.AppDomainAppVirtualPath.EndsWith("/") ? "" : "/")
                    + "webeditor.ashx",
                Query = "type=csv"
            };

            // create a mail merge config
            var mailMergeConfig = new Dictionary<string, object>
            {
                { "fileType", "csv" },
                { "url", mailMergeUrl.ToString()}
            };

            if (IsEnabledDirectUrl)
            {
                mailMergeConfig.Add("directUrl", directMailMergeUrl.ToString());
            }

            if (JwtManager.Enabled)  // if the secret key to generate token exists
            {
                var mailmergeToken = JwtManager.Encode(mailMergeConfig);  // encode mailMergeConfig into the token
                mailMergeConfig.Add("token", mailmergeToken);  // and add it to the mail merge config
            }

            dataSpreadsheet = jss.Serialize(mailMergeConfig);
        }

        //get a users for mentions
        public void GetUsersMentions(HttpRequest request, out string usersForMentions)
        {
            var jss = new JavaScriptSerializer();
            var id = request.Cookies.GetOrDefault("uid", null);
            var user = Users.getUser(id);
            usersForMentions = !user.id.Equals("uid-0") ? jss.Serialize(Users.getUsersForMentions(user.id)) : null;
        }
    }
}

File: web/documentserver-example/ruby/app/models/file_model.rb
# frozen_string_literal: true

#
# (c) Copyright Ascensio System SIA 2025
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require_relative '../configuration/configuration'

# Class for handling file-related operations and information.
class FileModel
  attr_accessor :file_name
  attr_accessor :mode
  attr_accessor :type
  attr_accessor :user_ip
  attr_accessor :lang
  attr_accessor :user
  attr_accessor :action_data
  attr_accessor :direct_url
  attr_reader :config_manager

  # set file parameters
  def initialize(attributes = {})
    @file_name = attributes[:file_name]
    @mode = attributes[:mode]
    @type = attributes[:type] || 'desktop' # the default platform type is desktop
    @user_ip = attributes[:user_ip]
    @lang = attributes[:lang]
    @user = attributes[:user]
    @action_data = attributes[:action_data]
    @direct_url = attributes[:direct_url]
    @config_manager = ConfigurationManager.new
  end

  # get file extension from its name
  def file_ext
    File.extname(@file_name).downcase
  end

  # get file url
  def file_uri
    DocumentHelper.get_file_uri(@file_name, true)
  end

  # get file uri for document server
  def file_uri_user
    if @config_manager.storage_path.absolute?
      "#{download_url}&dmode=emb"
    else
      DocumentHelper.get_file_uri(
        @file_name,
        false
      )
    end
  end

  # get document type from its name (word, cell or slide)
  def document_type
    FileUtility.get_file_type(@file_name)
  end

  # generate the document key value
  def key
    uri = "#{DocumentHelper.cur_user_host_address(nil)}/#{@file_name}" # get current user host address
    stat = File.mtime(DocumentHelper.storage_path(@file_name, nil)) # get the modification time of the given file
    ServiceConverter.generate_revision_id("#{uri}.#{stat}")
  end

  # get callback url
  def callback_url
    DocumentHelper.get_callback(@file_name)
  end

  # get url to the created file
  def create_url
    DocumentHelper.get_create_url(document_type)
  end

  # get url to download a file
  def download_url(is_server_url: true)
    DocumentHelper.get_download_url(@file_name, is_server_url:)
  end

  # get current user host address
  def cur_user_host_address
    DocumentHelper.cur_user_host_address(nil)
  end

  # get config parameters
  def config
    editors_mode = @mode || 'edit' # mode: view/edit/review/comment/fillForms/embedded
    can_edit = DocumentHelper.edited_exts.include?(file_ext) # check if the document can be edited
    if ((!can_edit && editors_mode.eql?('edit')) || editors_mode.eql?('fillForms')) &&
       DocumentHelper.fill_forms_exts.include?(file_ext)
      editors_mode = 'fillForms'
      can_edit = true
    end
    submit_form = !editors_mode.eql?('view') && @user.id.eql?('uid-1') # Submit form button state
    mode = can_edit && !editors_mode.eql?('view') ? 'edit' : 'view'
    # templates image url in the "From Template" section
    templates_image_url = DocumentHelper.get_template_image_url(document_type)
    templates = [
      {
        image: '',
        title: 'Blank',
        url: create_url
      },
      {
        image: templates_image_url,
        title: 'With sample content',
        url: "#{create_url}&sample=true"
      }
    ]
    unless @user.goback.nil?
      @user.goback[:url] = DocumentHelper.get_server_url(false)
    end

    config = {
      type:,
      documentType: document_type,
      document: {
        title: @file_name,
        url: download_url,
        directUrl: enable_direct_url? ? download_url(is_server_url: false) : '',
        fileType: file_ext.delete('.'),
        key:,
        info: {
          owner: 'Me',
          uploaded: Time.zone.now.to_s,
          favorite: @user.favorite
        },
        permissions: { # the permission for the document to be edited and downloaded or not
          comment: ['view', 'fillForms', 'embedded', 'blockcontent'].exclude?(editors_mode),
          copy: @user.denied_permissions.exclude?('copy'),
          download: @user.denied_permissions.exclude?('download'),
          edit: can_edit && ['edit', 'view', 'filter', 'blockcontent'].include?(editors_mode),
          print: @user.denied_permissions.exclude?('print'),
          fillForms: ['view', 'comment', 'blockcontent'].exclude?(editors_mode),
          modifyFilter: !editors_mode.eql?('filter'),
          modifyContentControl: !editors_mode.eql?('blockcontent'),
          review: can_edit && (editors_mode.eql?('edit') || editors_mode.eql?('review')),
          chat: !@user.id.eql?('uid-0'),
          reviewGroups: @user.review_groups,
          commentGroups: @user.comment_groups,
          userInfoGroups: @user.user_info_groups,
          protect: @user.denied_permissions.exclude?('protect')
        },
        referenceData: {
          instanceId: DocumentHelper.get_server_url(false),
          fileKey: unless @user.id.eql?('uid-0')
                     {
                       fileName: @file_name,
                       userAddress: DocumentHelper.cur_user_host_address(nil)
                     }.to_json
                   end
        }
      },
      editorConfig: {
        actionLink: @action_data ? JSON.parse(@action_data) : nil,
        mode:,
        lang: @lang || 'en',
        callbackUrl: callback_url, # absolute URL to the document storage service
        coEditing: if editors_mode.eql?('view') && @user.id.eql?('uid-0')
                     {
                       mode: 'strict',
                       change: false
                     }
                   end,
        createUrl: @user.id.eql?('uid-0') ? nil : create_url,
        templates: @user.templates ? templates : nil,
        user: { # the user currently viewing or editing the document
          id: @user.id.eql?('uid-0') ? nil : @user.id,
          name: @user.name,
          group: @user.group,
          image: @user.avatar ? "#{DocumentHelper.get_server_url(false)}/assets/#{@user.id}.png" : nil
        },
        embedded: { # the parameters for the embedded document type
          # the absolute URL that will allow the document to be saved onto the user personal computer
          saveUrl: download_url(is_server_url: false),
          # the absolute URL to the document serving as a source file for the document embedded into the web page
          embedUrl: download_url(is_server_url: false),
          # the absolute URL that will allow other users to share this document
          shareUrl: download_url(is_server_url: false),
          toolbarDocked: 'top' # the place for the embedded viewer toolbar (top or bottom)
        },
        customization: { # the parameters for the editor interface
          about: true, # the About section display
          comments: true,
          feedback: true, # the Feedback & Support menu button display
          forcesave: false, # adding the request for the forced file saving to the callback handler
          submitForm: submit_form, # the Submit form button state
          goback: @user.goback.nil? ? '' : @user.goback,
          close: @user.close.nil? ? '' : @user.close
        }
      }
    }

    if JwtHelper.enabled? # check if a secret key to generate token exists or not
      config['token'] = JwtHelper.encode(config) # encode a payload object into a token and write it to the config
    end

    config
  end

  # get document history
  def history
    file_name = @file_name
    file_ext = File.extname(file_name).downcase
    doc_key = key
    file_uri

    # get the path to the file history
    hist_dir = DocumentHelper.history_dir(DocumentHelper.storage_path(@file_name, nil))
    cur_ver = DocumentHelper.get_file_version(hist_dir) # get the file version

    if cur_ver.positive? # if file was modified
      hist = []
      hist_data = {}

      (1..cur_ver).each do |i| # run through all the file versions
        obj = {}
        data_obj = {}
        ver_dir = DocumentHelper.version_dir(hist_dir, i) # get the path to the given file version

        # get document key
        cur_key = doc_key
        if i != cur_ver
          File.open(File.join(ver_dir, 'key.txt'), 'r') do |file|
            cur_key = file.read
          end
        end
        obj['key'] = cur_key
        obj['version'] = i

        # check if the createdInfo.json file with meta data exists
        if (i == 1) && File.file?(File.join(hist_dir, 'createdInfo.json'))
          File.open(File.join(hist_dir, 'createdInfo.json'), 'r') do |file| # open it
            cr_info = JSON.parse(file.read) # parse the file content

            # write information about changes to the object
            obj['created'] = cr_info['created']
            obj['user'] = {
              id: cr_info['uid'],
              name: cr_info['uname']
            }
          end
        end

        # get the history data from the previous file version and write key and url information about it
        data_obj['fileType'] = file_ext[1..file_ext.length]
        data_obj['key'] = cur_key
        data_obj['url'] =
          if i == cur_ver
            DocumentHelper.get_download_url(
              file_name
            )
          else
            DocumentHelper.get_historypath_uri(
              file_name,
              i,
              "prev#{file_ext}"
            )
          end
        if enable_direct_url? == true
          data_obj['directUrl'] =
            if i == cur_ver
              download_url(is_server_url: false)
            else
              DocumentHelper.get_historypath_uri(
                file_name,
                i,
                "prev#{file_ext}",
                false
              )
            end
        end
        data_obj['version'] = i

        if i > 1 # check if the version number is greater than 1
          changes = nil
          # get the path to the changes.json file
          File.open(File.join(DocumentHelper.version_dir(hist_dir, i - 1), 'changes.json'), 'r') do |file|
            changes = JSON.parse(file.read) # and parse its content
          end

          change = changes['changes'].last

          # write information about changes to the object
          obj['changes'] = change ? changes['changes'] : nil
          obj['serverVersion'] = changes['serverVersion']
          obj['created'] = change ? change['created'] : nil
          obj['user'] = change ? change['user'] : nil

          prev = hist_data[(i - 2).to_s] # get the history data from the previous file version
          # write key and url information about previous file version with optional direct url
          data_obj['previous'] = if enable_direct_url? == true
                                   { # write key and url information about previous file version with optional directUrl
                                     fileType: prev['fileType'],
                                     key: prev['key'],
                                     url: prev['url'],
                                     directUrl: prev['directUrl']
                                   }
                                 else
                                   {
                                     fileType: prev['fileType'],
                                     key: prev['key'],
                                     url: prev['url']
                                   }
                                 end

          diff_path = [hist_dir, (i - 1).to_s, 'diff.zip'].join(File::SEPARATOR)
          if File.exist?(diff_path)
            # write the path to the diff.zip archive with differences in this file version
            data_obj['changesUrl'] = DocumentHelper.get_historypath_uri(file_name, i - 1, 'diff.zip')
          end
        end

        if JwtHelper.enabled? # check if a secret key to generate token exists or not
          # encode a payload object into a token and write it to the data object
          data_obj['token'] = JwtHelper.encode(data_obj)
        end

        hist.push(obj) # add object dictionary to the hist list
        hist_data[(i - 1).to_s] = data_obj # write data object information to the history data
      end

      return {
        hist: { # write history information about the current file version to the hist
          currentVersion: cur_ver,
          history: hist
        },
        histData: hist_data
      }
    end

    nil
  end

  # get image information
  def insert_image
    # image file type
    # server url to the image
    # direct url to the image
    insert_image = if enable_direct_url? == true
                     {
                       fileType: 'svg', # image file type
                       url: "#{DocumentHelper.get_server_url(true)}/assets/logo.svg", # server url to the image
                       directUrl: "#{DocumentHelper.get_server_url(false)}/assets/logo.svg" # direct url to the image
                     }
                   else
                     {
                       fileType: 'svg', # image file type
                       url: "#{DocumentHelper.get_server_url(true)}/assets/logo.svg" # server url to the image
                     }
                   end

    if JwtHelper.enabled? # check if a secret key to generate token exists or not
      # encode a payload object into a token and write it to the insert_image object
      insert_image['token'] = JwtHelper.encode(insert_image)
    end

    insert_image.to_json.tr('{', '').tr('}', '')
  end

  # get compared file information
  def data_document
    # file type
    # server url to the compared file
    # direct url to the compared file
    compare_file = if enable_direct_url? == true
                     {
                       fileType: 'docx', # file type
                       # server url to the compared file
                       url: "#{DocumentHelper.get_server_url(true)}/asset?fileName=sample.docx",
                       # direct url to the compared file
                       directUrl: "#{DocumentHelper.get_server_url(false)}/asset?fileName=sample.docx"
                     }
                   else
                     {
                       fileType: 'docx', # file type
                       # server url to the compared file
                       url: "#{DocumentHelper.get_server_url(true)}/asset?fileName=sample.docx"
                     }
                   end

    if JwtHelper.enabled? # check if a secret key to generate token exists or not
      # encode a payload object into a token and write it to the compare_file object
      compare_file['token'] = JwtHelper.encode(compare_file)
    end

    compare_file
  end

  # get mail merge recipients information
  def data_spreadsheet
    # file type
    # server url to the mail merge recipients file
    # direct url to the mail merge recipients file
    data_spreadsheet = if enable_direct_url? == true
                         {
                           fileType: 'csv', # file type
                           # server url to the mail merge recipients file
                           url: "#{DocumentHelper.get_server_url(true)}/csv",
                           # direct url to the mail merge recipients file
                           directUrl: "#{DocumentHelper.get_server_url(false)}/csv"
                         }
                       else
                         {
                           fileType: 'csv', # file type
                           # server url to the mail merge recipients file
                           url: "#{DocumentHelper.get_server_url(true)}/csv"
                         }
                       end

    if JwtHelper.enabled? # check if a secret key to generate token exists or not
      # encode a payload object into a token and write it to the data_spreadsheet object
      data_spreadsheet['token'] = JwtHelper.encode(data_spreadsheet)
    end

    data_spreadsheet
  end

  # get users data for mentions
  def users_mentions
    @user.id.eql?('uid-0') ? nil : Users.get_users_for_mentions(@user.id)
  end

  def users_info
    users_info = []
    return if @user.id.eql?('uid-0')

    Users.all_users.each do |user_info|
      u = {
        id: user_info.id,
        name: user_info.name,
        email: user_info.email,
        group: user_info.group,
        reviewGroups: user_info.review_groups,
        commentGroups: user_info.comment_groups,
        userInfoGroups: user_info.user_info_groups,
        favorite: user_info.favorite,
        deniedPermissions: user_info.denied_permissions,
        descriptions: user_info.descriptions,
        templates: user_info.templates,
        avatar: user_info.avatar
      }
      u['image'] = user_info.avatar ? "#{DocumentHelper.get_server_url(false)}/assets/#{user_info.id}.png" : nil
      users_info.push(u)
    end
    users_info
  end

  # get users data for protect
  def users_protect
    @user.id.eql?('uid-0') ? nil : Users.get_users_for_protect(@user.id)
  end

  # get direct url existence flag
  def enable_direct_url?
    !@direct_url.nil? && @direct_url == 'true'
  end
end
